File: app/src/main/java/com/blues/module_gank/_new/mvp/view/GankActivity.java
package com.blues.module_gank._new.mvp.view;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.Gravity;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.blues.R;
import com.blues.WebViewDialog;
import com.blues.framework.base.BaseActivity;
import com.blues.framework.base.BaseViewPagerAdapter;
import com.blues.framework.widget.endlessbannerview.BannerAdapter;
import com.blues.framework.widget.endlessbannerview.BannerView;
import com.blues.module_gank._new.mvp.model.GankBannerEntity;
import com.blues.module_gank._new.mvp.model.GankCategory;
import com.blues.module_gank._new.mvp.GankContract;
import com.blues.module_gank._new.mvp.model.GankUniversalEntity;
import com.blues.module_gank._new.detail.GankPostComments;
import com.blues.module_gank._new.detail.GankPostDetail;
import com.blues.module_gank._new.hot.GankHotEntity;
import com.blues.module_gank._new.mvp.presenter.GankPresenter;
import com.blues.module_gank._new.random.GankRandomEntity;
import com.blues.module_gank._new.search.GankSearchResult;
import com.google.android.material.tabs.TabLayout;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.ViewPager;

public class GankActivity extends BaseActivity implements GankContract.iGankView {

    private GankContract.iGankPresenter iGankPresenter;
    private BannerView mBanner;
    private TabLayout mTabLayout;
    private ViewPager mViewPager;

    //    private String[] tabs = {"Video", "Android", "iOS", "Web", "Blues"};
    private List<String> mTabs = new ArrayList<>();

    @Override
    public int setLayoutResourceId() {
        return R.layout.activity_gank;
    }

    @Override
    public void initView(Bundle savedInstanceState) {

        mPresenter = new GankPresenter(this);

        iGankPresenter.initData();

        mTabLayout = findViewById(R.id.tl_gank_top);
        mBanner = findViewById(R.id.bv_header);
        mViewPager = findViewById(R.id.vp_show);
    }

    private View getTabView(Context context, String text) {
        TextView textView = new TextView(context);
        textView.setGravity(Gravity.CENTER);
        textView.setText(text);
        return textView;
    }

    @Override
    public void setBanner(List<GankBannerEntity.DataBean> banner) {
        List<String> bannerUrls = new ArrayList<>();
        List<String> pathUrls = new ArrayList<>();
        for (GankBannerEntity.DataBean data : banner) {
            bannerUrls.add(data.getImage());
            pathUrls.add(data.getUrl());
        }
        startBannerLoop(bannerUrls, pathUrls);
    }

    @Override
    public void setCategory(List<GankCategory.DataBean> category) {
        for (GankCategory.DataBean data : category) {
            mTabs.add(data.getTitle());
        }

        BaseViewPagerAdapter mFragmentAdapter = new BaseViewPagerAdapter(mContext.getSupportFragmentManager(), mTabs);
        for (GankCategory.DataBean data : category) {
            mFragmentAdapter.addFragment(new GankUniversalFragment(data.getType()));
        }
        mViewPager.setAdapter(mFragmentAdapter);
        mViewPager.setOffscreenPageLimit(1);

        mViewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(mTabLayout));

        mTabLayout.addOnTabSelectedListener(new TabLayout.ViewPagerOnTabSelectedListener(mViewPager) {
            @Override
            public void onTabSelected(@NonNull TabLayout.Tab tab) {
                mViewPager.setCurrentItem(tab.getPosition());

                View view = tab.getCustomView();
                if (view instanceof TextView) {
                    ((TextView) view).setTextColor(view.getResources().getColor(R.color.color_light_blue));
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                View view = tab.getCustomView();
                if (view instanceof TextView) {
                    ((TextView) view).setTextColor(view.getResources().getColor(R.color.color_black));
                }
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });

        mTabLayout.setSelectedTabIndicatorColor(mTabLayout.getResources().getColor(R.color.color_light_blue));
        //因为fragment不销毁，所以添加新tabs前最好把旧的都移除掉先，否则就会出现重复的
        mTabLayout.removeAllTabs();

        for (int i = 0; i < mTabs.size(); i++) {
            mTabLayout.addTab(mTabLayout.newTab().setCustomView(getTabView(mContext, mTabs.get(i))), i == 0);
        }
    }

    @Override
    public void setPostList(List<GankUniversalEntity.DataBean> postList) {

    }

    @Override
    public void setRandomPostList(List<GankRandomEntity.DataBean> randomPostList) {

    }

    @Override
    public void setPostDetail(GankPostDetail.DataBean postDetail) {

    }

    @Override
    public void setPostHotList(List<GankHotEntity.DataBean> hotList) {

    }

    @Override
    public void setPostComments(GankPostComments comments) {

    }

    @Override
    public void setSearchContents(List<GankSearchResult.DataBean> searchContents) {

    }

    @Override
    public void setPresenter(GankContract.iGankPresenter presenter) {
        iGankPresenter = presenter;
    }

    private void startBannerLoop(List<String> images, List<String> urls) {
        new Handler(Looper.getMainLooper()).post(() -> {
            mBanner.setPlayDelay(3000);
            //启用progressbar形式hintview
//            bannerView.setHintView(new ProgressHintView(ctx, 0xffffffff, 0xff505050, 3000));
            mBanner.setHintPadding(mBanner.dp2px(10), 0, mBanner.dp2px(10), mBanner.dp2px(10));
            mBanner.setAnimationDuration(1000);
            LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, mBanner.dp2px(185));
//            layoutParams.setMargins(bannerView.dp2px(15f), bannerView.dp2px(10f), bannerView.dp2px(15f), 0);
            mBanner.setLayoutParams(layoutParams);
            BannerAdapter adapter = new BannerAdapter(mContext, images, mBanner);
            adapter.setItemClickListener(realPosition -> new WebViewDialog(mContext, urls.get(realPosition)).show());
            //开启轮播
            mBanner.setAdapter(adapter);
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mBanner != null) {
            mBanner.resume();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mBanner != null) {
            mBanner.pause();
        }
    }
}

File: app/src/main/java/com/blues/gankio/v2/view/GankActivity.kt
package com.blues.gankio.v2.view

import android.content.Context
import com.blues.framework.widget.endlessbannerview.BannerView
import com.google.android.material.tabs.TabLayout
import androidx.viewpager.widget.ViewPager
import com.blues.R
import android.os.Bundle
import android.widget.TextView
import android.view.Gravity
import com.blues.framework.base.BaseViewPagerAdapter
import com.google.android.material.tabs.TabLayout.TabLayoutOnPageChangeListener
import android.view.View
import android.widget.LinearLayout
import androidx.lifecycle.lifecycleScope
import com.blues.framework.widget.endlessbannerview.BannerAdapter
import com.blues.framework.dialog.WebViewDialog
import com.blues.framework.base.BaseKoinActivity
import com.blues.framework.utils.dp
import com.blues.gankio.v2.vm.GankioViewModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel
import java.util.ArrayList

class GankActivity : BaseKoinActivity() {

    private val gankioViewModel: GankioViewModel by viewModel()

    private val mBanner: BannerView by lazy {
        findViewById(R.id.bv_header)
    }
    private val mTabLayout: TabLayout by lazy {
        findViewById(R.id.tl_gank_top)
    }
    private val mViewPager: ViewPager by lazy {
        findViewById(R.id.vp_show)
    }

    private val mTabs: MutableList<String> = ArrayList()

    private fun getTabView(context: Context, textStr: String): View = TextView(context).apply {
        gravity = Gravity.CENTER
        text = textStr
    }

    override fun collect() {
        lifecycleScope.launch {
            gankioViewModel.banner.collect { bean ->
                startBannerLoop(bean.data.map { it.image } as MutableList<String>,
                    bean.data.map { it.url } as MutableList<String>)
            }
        }

        lifecycleScope.launch {
            gankioViewModel.category.collect {

                val mFragmentAdapter = BaseViewPagerAdapter(supportFragmentManager, mTabs)

                it.data.forEach { data ->
                    mTabs.add(data.title)

                    mFragmentAdapter.addFragment(GankUniversalFragment(data.type))
                }

                mViewPager.apply {
                    adapter = mFragmentAdapter
                    offscreenPageLimit = 1
                    addOnPageChangeListener(TabLayoutOnPageChangeListener(mTabLayout))
                }
                mTabLayout.addOnTabSelectedListener(object :
                    TabLayout.ViewPagerOnTabSelectedListener(mViewPager) {
                    override fun onTabSelected(tab: TabLayout.Tab) {
                        mViewPager.currentItem = tab.position

                        tab.customView.takeIf { view ->
                            view is TextView
                        }?.let { view ->
                            (view as TextView).setTextColor(
                                view.getResources().getColor(R.color.color_light_blue, null)
                            )
                        }
                    }

                    override fun onTabUnselected(tab: TabLayout.Tab) {

                        tab.customView.takeIf { view ->
                            view is TextView
                        }?.let { view ->
                            (view as TextView).setTextColor(
                                view.getResources().getColor(R.color.color_black, null)
                            )
                        }
                    }

                    override fun onTabReselected(tab: TabLayout.Tab) {}
                })

                with(mTabLayout) {
                    setSelectedTabIndicatorColor(
                        resources.getColor(
                            R.color.color_light_blue,
                            null
                        )
                    )

                    //因为fragment不销毁，所以添加新tabs前最好把旧的都移除掉先，否则就会出现重复的
                    removeAllTabs()

                    for (i in mTabs.indices) {
                        addTab(
                            newTab().setCustomView(getTabView(this@GankActivity, mTabs[i])),
                            i == 0
                        )
                    }
                }
            }
        }

    }

    private fun startBannerLoop(images: MutableList<String>, urls: MutableList<String>) {
        CoroutineScope(Dispatchers.Main).launch {
            val lp = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                185f.dp.toInt()
            )
            mBanner.apply {
                setPlayDelay(3000)
                setHintPadding(10f.dp.toInt(), 0, 10f.dp.toInt(), 10f.dp.toInt())
                setAnimationDuration(1000)
                layoutParams = lp
            }
            BannerAdapter(this@GankActivity, images, mBanner).apply {
                setItemClickListener(object : BannerAdapter.ItemClickListener {
                    override fun onClick(realPosition: Int) {
                        WebViewDialog(this@GankActivity, urls[realPosition]).show()
                    }
                })
            }.also { //开启轮播
                mBanner.setAdapter(it)
            }
        }
    }

    override fun getLayoutId(): Int = R.layout.activity_gank

    override fun initData(savedInstanceState: Bundle?) {
        gankioViewModel.requestData()
    }

    override fun onResume() {
        super.onResume()
        mBanner.resume()
    }

    override fun onPause() {
        super.onPause()
        mBanner.pause()
    }
}